core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -656,6 +656,69 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Nothing, O, Unit]
   def metered[F2[x] >: F[x]: Timer](rate: FiniteDuration): Stream[F2, O] =
     Stream.fixedRate[F2](rate).zipRight(this)
 
+  /**
+    * Logs the elements of this stream as they are pulled.
+    *
+    * If a tag is specified, it prefixes each element using the format `$tag: $o`.
+    * By default, log output is sent to standard out. Supply a different logger param
+    * to change this.
+    *
+    * This method does not change the chunk structure of the stream. To debug the
+    * chunk structure, see [[debugChunks]].
+    *
+    * Logging is not done in `F` because this operation is intended for debugging,
+    * including pure streams.
+    *
+    * @example {{{
+    * scala> Stream(1, 2).append(Stream(3, 4)).debug("a").toList
+    * a: 1
+    * a: 2
+    * a: 3
+    * a: 4
+    * res0: List[Int] = List(1, 2, 3, 4)
+    * }}}
+    */
+  def debug(tag: String = "", logger: String => Unit = println(_)): Stream[F, O] =
```

Comment:
Maybe unsafeLogger?

Reply:
Hm, the call site seems clunkier then -- `s.debug(unsafeLogger = println)`

Reply:
Just thinking about side effects in the API. Arguably `debug` itself suggests that there will be something shady going on (e.g. `toString` for formatting),  so maybe `logger` will also be an acceptable trade-off :)